### PR TITLE
Implemented encrypted passwords

### PR DESCRIPTION
**Pull request by georgesoteriou**

### CODE REVIEW

File: auth/auth.js [hunk 1]
```diff
@@ -12,7 +13,7 @@ exports.authSponsor = (user, pass, db, session, callback) => {
     if (err) {
       return logger.error('Unable to find sponsor: ' + err)
     } else {
-      if (result[0] && result[0].password === pass) {
+      if (result[0] && result[0].password === sha256(pass)) {
```
Comment (georgesoteriou) [inline auth/auth.js]: hashing input and comparing to already hashed password in db.

File: src/admin.js [hunk 2]
```diff
@@ -52,7 +53,7 @@ exports.setup = (app, db) => {
     // make sponsor
     var sponsor = new db.Sponsor({
       username: req.body.user,
-      password: req.body.pass,
+      password: sha256(req.body.pass),
```
Comment (georgesoteriou) [inline src/admin.js]: when creating sponsors, hashed password is saved

File: src/admin.js [hunk 3]
```diff
@@ -79,23 +80,47 @@ exports.setup = (app, db) => {
...
-      sponsor[0].password = req.body['s.password']
-      if (req.body['s.info.name']) { sponsor[0].info.name = req.body['s.info.name'] }
-      if (req.body['s.info.email']) { sponsor[0].info.email = req.body['s.info.email'] }
-      if (req.body['s.info.rank']) { sponsor[0].info.rank = req.body['s.info.rank'] }
-      if (req.body['s.info.bespoke']) { sponsor[0].info.bespoke = (req.body['s.info.bespoke'] === 'true') }
-      if (req.body['s.info.description']) { sponsor[0].info.description = req.body['s.info.description'] }
-      if (req.body['s.info.picture']) { sponsor[0].info.picture = req.body['s.info.picture'] }
-      if (req.body['s.info.link']) { sponsor[0].info.link = req.body['s.info.link'] }
+      if (req.body['s.password']) {
+        sponsor[0].password = sha256(req.body['s.password'])
```
Comment (georgesoteriou) [inline src/admin.js]: We can change the password by adding a new password in the admin table and a hash of the new password is save to db. Is this a good way to reset password if sponsor forgets it? do we need anything more really?

File: src/sponsor.js [hunk 4]
```diff
@@ -292,8 +294,8 @@ exports.setup = (app, db) => {
         logger.error('Failed to find sponsor: ' + err)
         return
       }
-      if (req.body.new && sponsor[0].password === req.body.old && req.body.new === req.body.new2) {
-        sponsor[0].password = req.body.new
+      if (req.body.new && sponsor[0].password === sha256(req.body.old) && req.body.new === req.body.new2) {
+        sponsor[0].password = sha256(req.body.new)
```
Comment (georgesoteriou) [inline src/sponsor.js]: when sponsor changes password, implemented hashing

File: src/admin.js (hunk 3 above)
Comment (georgesoteriou) [reply src/admin.js]: @jackel119

File: src/admin.js [hunk 5]
```diff
@@ -109,7 +134,9 @@ exports.setup = (app, db) => {
 
   // Remove Sponsor
   app.post('/admin/remove-sponsor/:user', (req, res) => {
-    db.Sponsor.remove({ username: req.params.user }, (err) => {
+    db.Sponsor.remove({
+      username: req.params.user
+    }, (err) => {
       if (err) {
```
Comment (jackpordi) [inline src/admin.js]: Should we do smth here?

**Review by jackpordi — DISMISSED** [review]
Would also say, change some of the fields from 'password' to 'password_hash' or 'pw_hash' to make it clear.

**Comment by plietar** [comment]
SHA256 for passwords is a really bad idea, and it seems like the hash isn't even salted. This makes it trivial to recover passwords from the hash, using rainbow tables.

For passwords you should be using salted PBKDF2 instead.

**Comment by jackpordi** [comment]
Good point.

@georgesoteriou use [this](https://www.npmjs.com/package/bcrypt) (bcrypt @plietar which is more secure) instead.

Use the async version with promises or async/await if you can.

File: src/admin.js (hunk 5 above)
Comment (georgesoteriou) [reply src/admin.js]: No, this should be fine

**Comment by georgesoteriou** [comment]
Ill work on it soon

File: auth/auth.js [hunk 6]
```diff
@@ -4,30 +4,41 @@ const rp = require('request-promise')
 const fs = require('fs')
 const yaml = require('js-yaml')
 const logger = require('../src/logger.js')
+var bcrypt = require('bcrypt');
```
Comment (jackpordi) [inline auth/auth.js]: any reason we're using var instead of const?

File: views/admin.pug [hunk 7]
```diff
@@ -1,5 +1,9 @@
 include ./layout.pug
 h1 Hello there DoCSoc
+
```
Comment (jackpordi) [inline views/admin.pug:3]: I assume these are just changes for the admin console?

Comment (georgesoteriou) [reply views/admin.pug:3]: yea so we can change sponsor password in case they forget it

**Review by jackpordi — APPROVED** [review]

**Pull request merged**